src/main/scala/viper/gobra/frontend/info/implementation/typing/ExprTyping.scala

```diff
@@ -235,8 +239,14 @@ trait ExprTyping extends BaseTyping { this: TypeInfoImpl =>
         ((n, exprType(n.left), exprType(n.right)) match {
           case (_: PEquals | _: PUnequals, l, r) => comparableTypes.errors(l, r)(n)
           case (_: PAnd | _: POr, l, r) => assignableTo.errors(l, AssertionT)(n) ++ assignableTo.errors(r, AssertionT)(n)
-          case (_: PLess | _: PAtMost | _: PGreater | _: PAtLeast | _: PAdd | _: PSub | _: PMul | _: PMod | _: PDiv, l, r) =>
+          case (_: PLess | _: PAtMost | _: PGreater | _: PAtLeast, l, r) =>
             assignableTo.errors(l, IntT(UntypedConst))(n) ++ assignableTo.errors(r, IntT(UntypedConst))(n)
+          case (_: PAdd | _: PSub | _: PMul | _: PMod | _: PDiv, l, r) =>
+            assignableTo.errors(l, IntT(UntypedConst))(n) ++ assignableTo.errors(r, IntT(UntypedConst))(n) ++ {
+              val constVal = intConstantEval(n)
+              val typCtx = getTypeFromContext(n.asInstanceOf[PNumExpression])
+              if (constVal.isDefined && typCtx.isDefined) assignableWithinBounds.errors(typCtx.get, n)(n) else noMessages
```

Comment:
A minor thing. You should first check wether constVal is defined and then comput the expected type (to avoid unnecessary computations).

Reply:
(Just as scala practice, do whathever you prefer) You could write the following:
```
val res = for {
  constVal <- intConstantEval(n)
  typeCtx <- getTypeFromContext(n.asInstanceOf[PNumExpression])
} yield assignableWithinBounds.errors(typCtx.get, n)(n)
res.getOrElse(noMessages)
```

Reply:
I realised that I don't even need to compute the value of the expression here, as this is made in `assignableWithinBounds`. As such, I could probably compress the code even further by writing
```scala
getTypeFromContext(n.asInstanceOf[PNumExpression]).map(assignableWithinBounds.errors(_, n)(n)).getOrElse(noMessages)
```
but this is hard to read so I will adapt your suggestion